File: micronautFactorial/micronaut-factorial/src/main/java/com/synclab/internship/FactorialService.java
package com.synclab.internship;

import javax.inject.Singleton;
import java.math.BigInteger;
import java.util.concurrent.locks.ReentrantLock;

@Singleton
public class FactorialService {

    //private static ReentrantLock reentrantLock = new ReentrantLock(true);

    public BigInteger factorial(int number) {
        //reentrantLock.lock();
        try {
            BigInteger result = new BigInteger("1");
            while (number > 1) {
                result = result.multiply(BigInteger.valueOf(number));
                number--;
            }
            return result;
        } finally {
            //reentrantLock.unlock();
        }
    }
}

File: ktor-factorial/src/FactorialService.kt
package com.synclab.internship

import java.math.BigInteger

class FactorialService {

    fun factorial(number: Int?): BigInteger? {
        if (number != null) {
            var result = BigInteger("1")
            for (i in number downTo 2) {
                result = result.multiply(BigInteger.valueOf(i.toLong()))
            }
            return result
        }
        return null
    }
}